File: Easy/Count distinct elements in every window/count-distinct-elements-in-every-window.py

class Solution:
    def countDistinct(self, A, N, K):
        ans=[]
        d={}
        for i in range(K):
            d[A[i]]=d.get(A[i],0)+1
        ans.append(len(d))
        for i in range(K,N):
            d[A[i-k]]=d.get(A[i-k],0)-1
            if d[A[i-k]]==0:
                del d[A[i-k]]
            d[A[i]]=d.get(A[i],0)+1
            ans.append(len(d))
        return ans
            
            


#{ 
 # Driver Code Starts
if __name__=='__main__':
    t = int(input())
    for i in range(t):
        n,k = list(map(int, input().strip().split()))
        arr = list(map(int, input().strip().split()))
        res = Solution().countDistinct(arr, n, k)
        for i in res:
            print (i, end = " ")
        print ()
# } Driver Code Ends

File: Difficulty: Easy/Count distinct elements in every window/count-distinct-elements-in-every-window.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.util.HashMap;

class GFG
{
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t > 0)
        {
            int n = sc.nextInt();
            int k = sc.nextInt();
            int a[] = new int[n];
            for (int i = 0; i < n; i++) 
                a[i] = sc.nextInt();
            Solution g = new Solution();
            
            ArrayList<Integer> ans = g.countDistinct(a, n, k);

            for (Integer val: ans) 
                System.out.print(val+" "); 
            System.out.println();
            t--;
        }
    }
}
// } Driver Code Ends


class Solution
{
    ArrayList<Integer> countDistinct(int A[], int n, int k){
        ArrayList<Integer> ans=new ArrayList<>();
        HashMap<Integer,Integer> countMap=new HashMap<>();
        for(int i=0;i<k;i++){
            countMap.put(A[i],countMap.getOrDefault(A[i],0)+1);
        }
        ans.add(countMap.size());
        for(int i=k;i<n;i++){
            countMap.put(A[i-k],countMap.get(A[i-k])-1);
            if(countMap.get(A[i-k])==0){
                countMap.remove(A[i-k]);
            }
            countMap.put(A[i],countMap.getOrDefault(A[i],0)+1);
            ans.add(countMap.size());
        }
        return ans;
    }
}

